rclrs/src/parameter/service.rs

```diff
@@ -1,89 +1,264 @@
+                // TODO(luca) error logging if the service call failed
+                DescribeParameters_Response { descriptors }
             },
         )?;
-        let get_parameter_types_service = Service::new(Arc::clone(&rcl_node_mtx),
-            "get_parameter_types",
-            |req_id: &rmw_request_id_t, req: GetParameterTypes_Request| {
-                GetParameterTypes_Response {
-                    types: seq![]
-                }
+        let map = parameter_map.clone();
+        let get_parameter_types_service = node.create_service(
+            &(fqn.clone() + "/get_parameter_types"),
+            move |_req_id: &rmw_request_id_t, req: GetParameterTypes_Request| {
+                let map = map.lock().unwrap();
+                let types = req
+                    .names
+                    .into_iter()
+                    .map(|name| {
+                        let name = name.to_cstr().to_str().ok()?;
+                        map.storage.get(name).map(|s| s.to_parameter_type())
+                    })
+                    .collect::<Option<_>>()
+                    .unwrap_or_default();
+                GetParameterTypes_Response { types }
             },
         )?;
-        let get_parameters_service = Service::new(Arc::clone(&rcl_node_mtx),
-            "get_parameters",
-            |req_id: &rmw_request_id_t, req: GetParameters_Request| {
-                GetParameters_Response {
-                    values: seq![]
-                }
+        let map = parameter_map.clone();
+        let get_parameters_service = node.create_service(
+            &(fqn.clone() + "/get_parameters"),
+            move |_req_id: &rmw_request_id_t, req: GetParameters_Request| {
+                let map = map.lock().unwrap();
+                let values = req
+                    .names
+                    .into_iter()
+                    .map(|name| {
+                        let name = name.to_cstr().to_str().ok()?;
+                        match map.storage.get(name)? {
+                            ParameterStorage::Declared(storage) => match &storage.value {
+                                DeclaredValue::Mandatory(v) => {
+                                    Some(v.read().unwrap().clone().into())
+                                }
+                                DeclaredValue::Optional(v) => Some(
+                                    v.read()
+                                        .unwrap()
+                                        .clone()
+                                        .map(|v| v.into())
+                                        .unwrap_or_default(),
+                                ),
+                                DeclaredValue::ReadOnly(v) => Some(v.clone().into()),
+                            },
+                            ParameterStorage::Undeclared(value) => Some(value.clone().into()),
+                        }
+                    })
+                    .collect::<Option<_>>()
```

Comment:
Same concern about collection here. Unknown parameter names should return a value of unset.

Reply:
Same behavior as the others:

```
lucadv@ros:~$ ros2 service call /minimal_publisher/get_parameters rcl_interfaces/srv/GetParameters "{names: [use_sim_time]}"                 
requester: making request: rcl_interfaces.srv.GetParameters_Request(names=['use_sim_time'])

response:
rcl_interfaces.srv.GetParameters_Response(values=[rcl_interfaces.msg.ParameterValue(type=1, bool_value=False, integer_value=0, double_value=0.0, string_value='', byte_array_value=[], bool_array_value=[], integer_array_value=[], double_array_value=[], string_array_value=[])])

lucadv@ros:~$ ros2 service call /minimal_publisher/get_parameters rcl_interfaces/srv/GetParameters "{names: [use_sim_time, hello]}"
requester: making request: rcl_interfaces.srv.GetParameters_Request(names=['use_sim_time', 'hello'])

response:
rcl_interfaces.srv.GetParameters_Response(values=[])
```